### ISSUE TITLE
Terms output in a strange order when paging in the Terms admin interface

### ISSUE THREAD

**Issue by pilotbean** [OrchardCMS/Orchard#7055]
Terms output in a strange order when paging in the Terms admin interface.

I believe this is because sort is applied after paging is applied.

https://github.com/OrchardCMS/Orchard/commit/356e8230ab2c4704ab300fb6915606137873bf22

**Comment by pilotbean** [comment]
Replacing line 45 in TermAdminController.cs with the following code fixes the issue:

```
            var terms = TermPart.Sort(_taxonomyService.GetTermsQuery(taxonomyId).ForPart<TermPart>().List());

            if (pager.PageSize > 0)
            {
                terms = terms.Skip(pager.GetStartIndex()).Take(pager.PageSize);
            }
```

This applies the paging after the sort.

**Comment by sebastienros** [comment]
Your change is just adding another issue, as it would load all the terms in memory before sorting and paging it.
However we agree there is an issue here, first the paging is not based on an order, so loading different pages doesn't mean we'll get the correct page, and also different hierarchies can be mixed in the same page.

The solution is not simple, but there is one. We need to change the FullPath property to have its components padded. Instead of have "1/2/3" we'd have "00000001/00000002/00000003" this way it is lexicographically orderable in database.

1- Change the code to store it this way.
2- Add an index to this field as we'll sort it in database, and also because we filter by prefix to load children of a specific term. Add the sort order in the paged query of the terms.
3- Make TermsComparer obsolete so that we can force users to use the database sorting.
4- Create an upgrade step in the Update module to update all existing terms.

**Comment by jersiovic** [comment]
Wouldn't it be better to add a new int property to taxonomy termpart called "Level".
Its value will be the number of '/' in FullPath. In migrations we should set it for existing terms.
Then the order will be obtained ordering by Level and Weight.

This is how we have solved this issue recently. Furthermore it helped us to add methods that get child elements of a node limiting the number of levels returned in an efficient way.
Don't you think it is more elegant and efficient than using padding?

**Comment by jersiovic** [comment]
Problem of that solution is ignores the weight of the terms. Other more complex solution would be to have a FullWeigthPath Property that stores weights of parents: instead of its id: "2/5/2".

The extra complexity would be caused because after update the weight of a term we will need to visit all their children updating the FullWeigthPath to update the related value.

**Comment by sebastienros** [comment]
So maybe we can keep the current Path, but create the lexicographically sortable structure using weights. Or like in menu items, use a `double` for weight that would just represent the order in the tree.

**Comment by jersiovic** [comment]
Problem I see is to update the double or the lexicographically sortable structure for all terms affected in an import scenario for each imported term is expensive. Unless we can execute the reorder operation for  the taxonomies affected only after a complete step of a recipe execution finished.

**Comment by pilotbean** [comment]
What size of dataset will see all these performance issues come into play
1000+ terms? 10,000+ terms?

Maybe just put warnings in place on relevant recipes/imports to say that
large sets of terms may take a long time to process.

There can't be that many people using datasets of this size. I think anyone
using really big term sets will be faced with a whole host of problems - I
mean how would the interface deal with selecting from 1000+ terms when
editing a content item?

Also with that number of terms surely the primary (and possibly only) sort
bring used would be alphabetical, which I assume doesn't have these issues,
as that is a sort naturally available already via DB. Sorting by weight
wouldn't have any practical use with sets this size.

So perhaps we make the weight feature something that is enabled/disabled
separately - so can be disabled for these large dataset scenarios.

**Comment by jersiovic** [comment]
Well for example we currently for this page https://www.sayandsail.es we use a taxonomy containing all the ports in Spain, a total of 467 ports categorized by country and province. As soon as we add new countries it is expected to grow up considerably. 
Currently importing it consumes considerable time.  In fact, in our daily develop cycle we create the site from scratch using a setup recipe but we avoid to create the complete taxonomy when it is not necessary for saving time. 
You are right we order this taxonomy terms setting the weight to values that follow an alphabetically order. But for ordering a hierarchy of terms alphabetically we still need to take into account the hierarchy, that means use the text of the parent of a term. So, we will need a lexicographically orderable structure based on the text of the parent terms: grandfather text/father text/my text which should be updated when the text of parent term changes. In our case this happened to correct text encoding mistakes in the original import we made to production site.

We have other multitenant configuration with 30 sites where each site can have around 360 terms of families of products localized in Spanish and English for a taxonomy. Every night those sites sync products and families from an ERP. Currently this operations takes more than it will be desirable. I know it can be improved on our side replacing current brute force sync per a syncing based on changed or removed families, but this another discussion about how to deal with legacy systems. The point is that in general orchard 1 is not very fast importing data so please don't make it slower.

**Comment by jersiovic** [comment]
But, I agree it would good to know the real cost of performing this operation for different numbers of terms to check if my concerns are well founded.

**Comment by jersiovic** [comment]
I think I didn't read with enough attention what @sebastienros said: "Or like in menu items, use a  double  for weight that would just represent the order in the tree."
This double won't need to be recalculated after update each term. It will be just the one you provide in the import. So, when I create the xml for the import I will take care of providing the right value.
So, I agree with that proposal. 👍

**Comment by jersiovic** [comment]
We are evaluating to provide a PR (me and @jardg) .
After looking how menuitems work I found they don't use double for weight https://github.com/OrchardCMS/Orchard/blob/dev/src/Orchard.Web/Core/Navigation/Controllers/AdminController.cs#L140
They use a string with position of ancestors and their own separated by dots. When the order of a parent is updated they update their children. They could be improved using a lexycographyc sortable position, but this is sth for another issue.
Coming back to taxonomies: 
- Working with a double for the weight will solve the issue of adding leading 0 to the position, but will force us to load and update all the terms of a taxonomy when one term is updated.
- Working with a string lexicographically sortable, will reduce the number of load and updates needed to the minimum possible only doing that for terms affected.
- Instead of using a handler for updating positions, as for menuitems, leaving the responsibility of providing the position to the code that adds or modifies a term allows a good performance for import scenarios, cause we don't need to set positions cause we rely on the position provided in the recipe. So, you need to provide in your recipe all the childrens affected by a change in position or to use a command that recalculate the position of those terms (or all of them) in a taxonomy.
- TaxonomyTerm driver should export positions without leading 0s and accept them without leading 0s

@sebastienros @pilotbean If do you agree with that we could work on a PR
